### PR TITLE
Transition Hebrew rvrn forms to compatible forms with opsz avar mapping support

### LINKED ISSUE #233: Missing perafe-hb.BRACKET.18 and samekhdagesh-hb.BRACKET.18 in Italic
@danielgrumer We seem to be missing these rvrn substitution opsz 18 glyphs in the Italic set only.  And they are not defined in the Italic set designspace rule set.  Intentional difference for these forms between Roman and Italic?

### PR DESCRIPTION
**Pull request by chrissimpkins**
Fixes https://github.com/googlefonts/googlesans/issues/233

Outline updates from Daniel Grumer.

This PR transitions all Hebrew rvrn @ opsz 18 outlines to compatible forms so that they work with our new avar mapping approach.  This allows us to eliminate the *.BRACKET.18 outline data associated with these forms.

With these changes there is no longer rvrn feature support in the project.  🥳 

String with the affected glyphs for testing:

ףףּפפּפֿססּ  

---

### TODO

- [x] Daniel Grumer review
- [x] remove commented out rvrn sub rules in the designspace files
- [x] update expected glyph name/order definitions with the elimination of the *.BRACKET.18 glyphs that are removed with these changes (https://github.com/googlefonts/googlesans/pull/281/commits/1ce3f323c18fc60ef22c6e0eae9dc22222152d19)
- [x] update fea tag checks to remove `rvrn` (all gone with the changes here) https://github.com/googlefonts/googlesans/pull/281/commits/d274d43aafece7a304e922aac711396f7f95d899
- [x] update shaping regression tests that include the now removed *.BRACKET.18 Hebrew glyph forms
    - [x] `ccmp_hebr` 6574ad54379276d95c5b603a319d2a7af84754cc
    - [x] `kern_hebr` c3bb2f07e518a5c1fc136463254bd34ec4dfd174
    - [x] review shaping changes
- [x] set all Hebrew .BRACKET.18 forms to non-export in the designspace files.
- [x] update feature files (see https://github.com/googlefonts/googlesans/pull/274#discussion_r610076563)
- [x] Run gs-normalizedesignspace.py to update the GDEF in feature.

### CODE REVIEW

**Comment by chrissimpkins** [comment]
@danielgrumer This seems to work but it looks like you moved away from the rounded top forms in the opsz 18 designs? 

Hebrew-rvrn

**Comment by chrissimpkins** [comment]
@madig Mind having a look at this next week?  Let me know if you have any concerns about the changes here and if there are any issues with the order of landing this in main vs. other open PR work in progress.  Ty!

File: qa/check-fea.py [hunk 1]
```diff
@@ -405,9 +403,7 @@ def com_google_fonts_check_googlesans_features_regression(ttFont, hb_font):
                         )
                         continue
                     if shaped_text != expected:
-                        shaped_texts_str = textwrap.indent(
-                            "\n".join(shaped_text), "\t  "
-                        )
+                        shaped_texts_str = textwrap.indent("\n".join(shaped_text), "\t  ")
```
Comment (chrissimpkins) [inline qa/check-fea.py:474]: Sorry, autoblackened by VS Code...

**Comment by chrissimpkins** [comment]
Daniel reached out to me and said that the *intent* is that the opsz 18 forms should maintain the rounded tops without the angled upper left corner. It sounds like these shapes are in the bracket layer glyph definitions and should be compatible with the opsz 17 masters.  I'll look into the contour swaps next week to see if we can make it work as he intends.

**Comment by madig** [comment]
I was about to say, the previous rvrn forms seem to be gone now. So... we can't remove rvrn here?

**Comment by chrissimpkins** [comment]
It sounds like they are still in the *.BRACKET.18 glyph definitions.  I *think* that we will need to do is move those outlines to the opsz 18 masters weight for weight.  This is per Daniel feedback though.  I haven't looked into it yet myself.  The rounded top shapes are shown in the OP image in https://github.com/googlefonts/googlesans/issues/233

**Comment by chrissimpkins** [comment]
My understanding is that they are now *compatible* but are simply in the wrong location for interpolation.

**Comment by madig** [comment]
Oh, ok. I was expecting Daniel to put the glyphs in the right place 😄 I'll try copying stuff over.

**Comment by chrissimpkins** [comment]
> Oh, ok. I was expecting Daniel to put the glyphs in the right place 😄 I'll try copying stuff over.

Likely my fault in the way it was communicated to him. If they are compatible it will be a simple swap.  I have the UFO formatted copies of his glyphs sources locally and will do this later in the week.  Once all pieces are in place and functioning I will poke you for a review so that we can make sure that this is ready to land.

**Comment by chrissimpkins** [comment]
> Oh, ok. I was expecting Daniel to put the glyphs in the right place 😄 I'll try copying stuff over.

Likely my fault in the way it was communicated to him. If they are compatible it will be a simple swap.  I have the UFO formatted copies of his glyphs sources locally and will do this later in the week.  Once all pieces are in place and functioning I will poke you for a review so that we can make sure that this is ready to land.

**Comment by danielgrumer** [comment]
@chrissimpkins I'm following this thread. let me know if you need help. 
Perhaps we can schedule a meeting for next week, and cancel it in case it all goes smooth for you?

**Comment by chrissimpkins** [comment]
> @chrissimpkins I'm following this thread. let me know if you need help. 
> 
> Perhaps we can schedule a meeting for next week, and cancel it in case it all goes smooth for you?

Thanks Daniel! I should be able to have a look tomorrow or Thursday.  Will be in touch once I understand what needs to be swapped and how it turns out.  Hopefully this sorts it out and no meeting is necessary but I would be happy to get together to run through the update or work through any problems that we encounter.  Will be in touch when I know more.

**Comment by chrissimpkins** [comment]
Found 'em!



Apparently the bright blue was too subtle for me... :)

**Comment by chrissimpkins** [comment]
@danielgrumer It doesn't appear that the glyphs are compatible between the opsz 17 and opsz 18 master .BRACKET.18 designs Daniel.  When I replace the default forms in opsz = 18 with the *.BRACKET.18 design counterpart in the opsz=18 masters, I am seeing incompatible issues like different point numbers in the samekh-hb contours shown here (opsz=17 master on the left, opsz=18 master on the right):



I'll prepare glyphs files for you with the current state of the merge and will need you to have a look at the opsz=17 to opsz=18 master compatibility across all of the following glyphs:

- pe-hb
- finalpe-hb
- samekh-hb

It looks like the others in this string of glyphs that currently use rvrn "ףףּפפּפֿססּ" are composites of the above forms. Once we address the compatibility of those three glyphs across the opsz 17 and opsz 18 masters, we should be ok.

**Comment by chrissimpkins** [comment]
Glyphs sources built @ commit https://github.com/googlefonts/googlesans/pull/281/commits/610e7bae68d42f69e25020d5c3a98ddea4d5c063

[GoogleSans-610e7ba.zip](https://github.com/googlefonts/googlesans/files/6246037/GoogleSans-610e7ba.zip)

**Comment by danielgrumer** [comment]
Hey @chrissimpkins 
Attached Glyphs files with these three letters, and the ones containing those components.
Marked in green.

Removed all the bracket layers. hope that now it will fit better!
Daniel

[To Chris.zip](https://github.com/googlefonts/googlesans/files/6254197/To.Chris.zip)

**Comment by chrissimpkins** [comment]
Thanks Daniel!

**Comment by chrissimpkins** [comment]
@danielgrumer Forms are compatible!  Nicely done.  I merged your changes in https://github.com/googlefonts/googlesans/pull/281/commits/4f60f313cf3f00b95252e55689223afd6cb8a463 and things look good to me.

Here is where we are with the full transition away from rvrn substitutions to compatible forms with opsz axis intermediate interpolated forms hidden by the avar table:

hebrew-rvrn-transition

Mind having a look at the Roman and Italic variable fonts in this archive to confirm that you are ok with these changes?

https://github.com/googlefonts/googlesans/suites/2442630236/artifacts/52299327

**Comment by chrissimpkins** [comment]
CI still fails after https://github.com/googlefonts/googlesans/pull/281/commits/4f60f313cf3f00b95252e55689223afd6cb8a463

It looks like the shaping definitions will require another update.  Will wait for Daniel to sign off and then look into this.

**Comment by chrissimpkins** [comment]
Daniel ok'd these changes.

**Comment by madig** [comment]
Rebased on top of current main.

**Comment by madig** [comment]
@googlebot I consent.

**Comment by madig** [comment]
Needed to update the shaping result files for some reason.

Edit: seems benign, no advance width actually changed.

**Review by madig — APPROVED** [review]

**Comment by madig** [comment]
@danielgrumer The dlig feature doesn't seem to do much for the vertical thingies? Intentional?

gshebr

**Comment by danielgrumer** [comment]
@madig 
No, the opposite. Actually the `dlig` feature is doing TOO MUCH... you found a small bug.
Out of what's in the screenshot, only the first line is supposed to look different. The 3rd line should look the same.

The reason this feature exists, is to support Yiddish precomposed ligatures. they don't look any different, but sits in its own unicode. If you want we can fix that.

**Comment by madig** [comment]
Cool 😬 hm, I'll try to finish going through this PR today, not sure if it's better to add a fix here or in a separate PR.

**Comment by danielgrumer** [comment]
This is a very small issue, imo it's better be fixed later.

Did you also have a question about the kern feature?
I saw it in my email, but can't track it down in GitHub.

**Comment by madig** [comment]
Yeah I deleted it, but actually I should post it: kerning doesn't seem to do all that much for the text we use in regression testing?
gshekern2

**Comment by madig** [comment]
I see no changes in the features, except the marked thing here moving to the left in the new files:

image

@danielgrumer

File: qa/definitions/GoogleSans[GRAD,opsz,wght].ttf.glyphsetdef [hunk 2]
```diff
@@ -3096,11 +3096,4 @@ bracketright.loclGURM
...
-quoteright.loclGURM
-uni05E3.BRACKET.18
-uniFB43.BRACKET.18
-uni05E4.BRACKET.18
-uniFB44.BRACKET.18
-uniFB4E.BRACKET.18
-uni05E1.BRACKET.18
-uniFB41.BRACKET.18
\ No newline at end of file
+quoteright.loclGURM
```
Comment (chrissimpkins) [inline qa/definitions/GoogleSans[GRAD,opsz,wght].ttf.glyphsetdef:3099]: No more `.BRACKET.18` is a beautiful thing.

File: source/GoogleSans/GoogleSans-opsz17-wght380-GRAD200.ufo/lib.plist [hunk 3]
```diff
@@ -8080,6 +8080,8 @@
       <string>f_j.alt</string>
       <string>f_k.alt</string>
       <string>f_t.alt</string>
+      <string>finalpe-hb.BRACKET.18</string>
```
Comment (chrissimpkins) [inline source/GoogleSans/GoogleSans-opsz17-wght380-GRAD200.ufo/lib.plist:8083]: Hmm.  Why are these still showing up in the lib.plist files?

File: source/GoogleSans/GoogleSans.designspace [hunk 4]
```diff
@@ -13,20 +13,6 @@
     </axis>
     <axis tag="GRAD" name="Grade" minimum="-50" maximum="200" default="0"/>
   </axes>
-  <rules>
```
Comment (chrissimpkins) [inline source/GoogleSans/GoogleSans.designspace:16]: 🥳

File: source/GoogleSans/GoogleSans.designspace [hunk 5]
```diff
@@ -510,6 +496,8 @@
         <string>f_j.alt</string>
         <string>f_k.alt</string>
         <string>f_t.alt</string>
+        <string>finalpe-hb.BRACKET.18</string>
```
Comment (chrissimpkins) [inline source/GoogleSans/GoogleSans.designspace:499]: I think I see what is going on.  These are being added to the non-export lists.  I wasn't getting that context in the GitHub diff

File: source/GoogleSans/GoogleSans-opsz17-wght380-GRAD200.ufo/lib.plist (hunk 3 above)
Comment (chrissimpkins) [reply source/GoogleSans/GoogleSans-opsz17-wght380-GRAD200.ufo/lib.plist:8083]: Resolved, these glyphs are being added to the non-export lists

File: qa/definitions/GoogleSans[GRAD,opsz,wght].ttf.glyphsetdef (hunk 2 above)
Comment (madig) [reply qa/definitions/GoogleSans[GRAD,opsz,wght].ttf.glyphsetdef:3099]: So much less headache.

**Comment by chrissimpkins** [comment]
@madig Mind opening an issue for the dlig issue that you identified.  IIUC Daniel feels that this is a bug

File: qa/definitions/GoogleSans[GRAD,opsz,wght].ttf.glyphsetdef (hunk 2 above)
Comment (chrissimpkins) [reply qa/definitions/GoogleSans[GRAD,opsz,wght].ttf.glyphsetdef:3099]: Until someone decides that opsz needs to expand and interpolate :)

**Comment by madig** [comment]
https://github.com/googlefonts/googlesans/issues/321

File: qa/definitions/GoogleSans[GRAD,opsz,wght].ttf.glyphsetdef (hunk 2 above)
Comment (madig) [reply qa/definitions/GoogleSans[GRAD,opsz,wght].ttf.glyphsetdef:3099]: Then we can offer to drop Adobe support.

**Comment by chrissimpkins** [comment]
Note to self: Always paste the test string into the PR thread.  Grr.  :)

Let me figure out what these glyphs are again so that I can have a look.

**Comment by chrissimpkins** [comment]
```
ססּףףּפפּפֿ
```

File: source/GoogleSans/GoogleSans-Italic.designspace [hunk 6]
```diff
@@ -13,18 +13,6 @@
     </axis>
     <axis tag="GRAD" name="Grade" minimum="-50" maximum="200" default="0"/>
   </axes>
-  <rules>
```
Comment (chrissimpkins) [inline source/GoogleSans/GoogleSans-Italic.designspace:16]: Marking the site of the [18] designspace rule removal for posterity

**Comment by chrissimpkins** [comment]
@madig Confirming that you ran the GDEF script on this update?  The box is checked but I know that I didn't do it and I don't see a commit message where this is indicated.  It looks like the GDEF defs *are* updated.

**Comment by chrissimpkins** [comment]
This looks great.

hebrew-rvrn

**Comment by chrissimpkins** [comment]
LGTM.  Let's merge 🥳 

Thanks again for all of your help with this @danielgrumer! Much appreciated.  We will push this update in our next release.

**Comment by chrissimpkins** [comment]
Changelog updated 01e330b14b3230f346d7ccb45264821750bdae6a

**Comment by madig** [comment]
> @madig Confirming that you ran the GDEF script on this update? The box is checked but I know that I didn't do it and I don't see a commit message where this is indicated. It looks like the GDEF defs _are_ updated.

I did indeed. Only the best for you.

**Pull request merged**